---
title: "assignment_3_4_ds"
output: html_document
date: '2022-03-23'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
suppressPackageStartupMessages({
  library(AnnotationHub)
  library(Rsubread)
  library(rtracklayer)
  library(Biostrings)
  library(Rfastp)
  library(epiwraps)
})
ah <- AnnotationHub()
```

## Using Rfastp

```{r}
dir.create("rfastp.trimmed")
qc <- Rfastp::rfastp("ENCFF001LJN.fastq.gz", outputFastq="rfastp.trimmed/p300", 
                     thread=4, overrepresentationAnalysis=TRUE)
```

```{r}
Rfastp::curvePlot(qc, curve="content_curves")
```
# Alignment

## Using Rsubread

### Building a genome index for mapping

This we have to do only once for a genome, and can then re-use across projects

```{r, eval=FALSE}
# we get the genome sequence from AnnotationHub for the mouse genome v102 as in week2 exercise.
#make sure to load the genome sequence as a two.bit file and not the EnsDB object!
genome <- ah[["AH14005"]]
# we create a new directory that will contain the genome index
dir.create("mouse_genome_index")
# we write the genome sequence in fasta format
export(import.2bit(genome), "mouse_genome_index/genome.fasta.gz", compress=TRUE)
# we build a Rsubread index
Rsubread::buildindex("mouse_genome_index/rsubread", reference="mouse_genome_index/genome.fasta.gz")
```



### Alignment

```{r, eval=FALSE}
dir.create("aligned")
align.stats <- Rsubread::align(index="mouse_genome_index/rsubread", type="dna",
                               output_file="aligned/p300.bam",
                               readfile1="rfastp.trimmed/p300_R1.fastq.gz", 
                               nthreads=6, sortReadsByCoordinates=TRUE)
align.stats
```

# Peak calling
## Using R
```{r, eval=FALSE}
peaks <- callPeaks("aligned/p300.bam")
# if we want to save it as a bed file:
rtracklayer::export.bed(peaks, "peaks/peaks.bed")
```


```{r, eval=FALSE}
ensdb <- ah[["AH89211"]] # Mouse (mm10) EnsDb
plotSignalTracks(c("aligned/p300.bam"), region="numb", type = "alignment", ensdb=ensdb, transcripts="full")
```

Since creating the mouse index on my local computer was a problem, I downloaded the specific traks for the ChIP experiment of p300 and various histone modifications locally to my computer and did the assignment like this. Above you could see the normal pipeline if we wanted to perform our own peak calling either in R or with MACS2 on a computing cluster in bash. 

```{r, eval=FALSE}
#How to read in the data directly from the source so the data will be always up to date and the script can be run independently of the local storage of the previously downloaded data. PL mentioned it in class. 
dir.create("peaks")
dir.create("tracks")
options(timeout=1800)
# p300
download.file("https://www.encodeproject.org/files/ENCFF460EOA/@@download/ENCFF460EOA.bed.gz", "peaks/p300.bed.gz")
download.file("https://www.encodeproject.org/files/ENCFF146SCZ/@@download/ENCFF146SCZ.bigWig", "tracks/p300.bw")
# H3K27ac
download.file("https://www.encodeproject.org/files/ENCFF274UIB/@@download/ENCFF274UIB.bed.gz", "peaks/H3K27ac.bed.gz")
download.file("https://www.encodeproject.org/files/ENCFF121THA/@@download/ENCFF121THA.bigWig", "tracks/H3K27ac.bw")
# H3K27me3
download.file("https://www.encodeproject.org/files/ENCFF008XKX/@@download/ENCFF008XKX.bed.gz", "peaks/H3K27me3.bed.gz")
download.file("https://www.encodeproject.org/files/ENCFF160FEV/@@download/ENCFF160FEV.bigWig", "tracks/H3K27me3.bw")
# H3K4me1
download.file("https://www.encodeproject.org/files/ENCFF333IJH/@@download/ENCFF333IJH.bed.gz", "peaks/H3K4me1.bed.gz")
download.file("https://www.encodeproject.org/files/ENCFF016YZA/@@download/ENCFF016YZA.bigWig", "tracks/H3K4me1.bw")
# H3K4me3
download.file("https://www.encodeproject.org/files/ENCFF247GVM/@@download/ENCFF247GVM.bed.gz", "peaks/H3K4me3.bed.gz")
download.file("https://www.encodeproject.org/files/ENCFF611GSQ/@@download/ENCFF611GSQ.bigWig", "tracks/H3K4me3.bw")
```


```{r}
#reading in the data
p300 <- import("peaks/p300.bed.gz", format="NarrowPeak")
k4me1 <- import("peaks/H3K4me1.bed.gz", format="NarrowPeak")
k4me3 <- import("peaks/H3K4me3.bed.gz", format="NarrowPeak")
k27ac <- import("peaks/H3K27ac.bed.gz", format="NarrowPeak")
k27me3 <- import("peaks/H3K27me3.bed.gz", format="NarrowPeak")
```

```{r}
#Here we compare the overlaps of the specific histone marks with the peaks of p300 data set.
k4me3 <- sum(overlapsAny(p300, k4me3))
k4me1 <- sum(overlapsAny(p300, k4me1))
k27ac <- sum(overlapsAny(p300, k27ac))
k27me3 <- sum(overlapsAny(p300, k27me3))

#to calculate the percentages:
#percentage of k4me3
k4me3/length(p300)
#percentage of k4me1
k4me1/length(p300)
#percentage of k27ac
k27ac/length(p300)
#percentage of k27me3
k27me3/length(p300)
```
